### PR TITLE
.github: Add intial Windows CPU GHA workflow

### PR DESCRIPTION
**Pull request by seemethere**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #58418 .github: Enable Windows workflow for pull_request
* **#58199 .github: Add intial Windows CPU GHA workflow**
* #58363 .github: Make on_pull_request a conditional block

Signed-off-by: Eli Uriegas <eliuriegas@fb.com>

Differential Revision: [D28465272](https://our.internmc.facebook.com/intern/diff/D28465272)

### CODE REVIEW

**Comment by seemethere** [comment]
@seemethere has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28465272).

**Comment by seemethere** [comment]
@seemethere has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28465272).

**Comment by seemethere** [comment]
@seemethere has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28465272).

File: .github/scripts/generate_ci_workflows.py [hunk 1]
```diff
@@ -8,11 +8,44 @@
...
 
+class PyTorchWindowsWorkflow:
+    CPU_TEST_RUNNER = "windows.4xlarge"
+    CUDA_TEST_RUNNER = "windows.8xlarge.nvidia.gpu"
+
+    def __init__(
+            self,
+            build_environment: str,
+            on_pull_request: bool = False
+    ):
```
Comment (malfet) [inline .github/scripts/generate_ci_workflows.py:20]: Nit
```suggestion
    ) -> None:
```

**Comment by samestep** [comment]
for anyone like me who wants to read only one of the pair of `.yml`/`.yml.in` files and not both, here's the output of `git diff --no-index .github/templates/windows_ci_workflow.yml.in .github/workflows/pytorch-win-vs2019-cpu-py3.yml`:
```diff
diff --git a/.github/templates/windows_ci_workflow.yml.in b/.github/workflows/pytorch-win-vs2019-cpu-py3.yml
index 0ba3f8110e..0fe5e87e1e 100644
--- a/.github/templates/windows_ci_workflow.yml.in
+++ b/.github/workflows/pytorch-win-vs2019-cpu-py3.yml
@@ -1,11 +1,9 @@
+# @generated DO NOT EDIT MANUALLY
 # Template is at:    .github/templates/windows_ci_workflow.yml
 # Generation script: .github/scripts/generate_ci_workflows.py
-name: Windows CI (!{{ build_environment }})
+name: Windows CI (pytorch-win-vs2019-cpu-py3)
 
 on:
-{%- if on_pull_request %}
-  pull_request:
-{%- endif %}
   push:
     branches:
       - master
@@ -13,7 +11,7 @@ on:
   workflow_dispatch:
 
 env:
-  BUILD_ENVIRONMENT: !{{ build_environment }}
+  BUILD_ENVIRONMENT: pytorch-win-vs2019-cpu-py3
   BUILD_WHEEL: 1
   CUDA_VERSION: "cpu"
   IN_CI: 1
@@ -53,9 +51,9 @@ jobs:
           name: ${{ env.BUILD_ENVIRONMENT }}
           path: C:\w\build-results
   test:
-    runs-on: !{{ test_runner_type }}
+    runs-on: windows.4xlarge
     env:
-      JOB_BASE_NAME: !{{ build_environment }}-test
+      JOB_BASE_NAME: pytorch-win-vs2019-cpu-py3-test
     needs:
       - build
     steps:
@@ -138,7 +136,7 @@ jobs:
           AWS_ACCESS_KEY_ID: ${{ secrets.AWS_OSSCI_METRICS_ACCESS_KEY_ID }}
           AWS_SECRET_ACCESS_KEY: ${{ secrets.AWS_OSSCI_METRICS_SECRET_ACCESS_KEY }}
           CIRCLE_BRANCH: ${{ steps.parse-ref.outputs.branch }}
-          CIRCLE_JOB: !{{ build_environment }}
+          CIRCLE_JOB: pytorch-win-vs2019-cpu-py3
           CIRCLE_PR_NUMBER: ${{ github.event.pull_request.number }}
           CIRCLE_SHA1: ${{ github.event.pull_request.head.sha || github.sha }}
           CIRCLE_TAG: ${{ steps.parse-ref.outputs.tag }}
```

File: .github/scripts/windows_install_utils.ps1 [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+        $filePath = "msiexec.exe"
+    }
+
+    try
+    {
+        Write-Host "Starting Install $Name..."
+        $process = Start-Process -FilePath $filePath -ArgumentList $ArgumentList -Wait -PassThru
+
+        $exitCode = $process.ExitCode
+        if ($exitCode -eq 0 -or $exitCode -eq 3010)
```
Comment (malfet) [inline .github/scripts/windows_install_utils.ps1:45]: Can you please add comment that `3010 == ERROR_SUCCESS_REBOOT_REQUIRED` (and perhaps a link to https://docs.microsoft.com/en-us/windows/win32/msi/error-codes

**Review by malfet — APPROVED** [review]

File: .github/scripts/windows_install_utils.ps1 (hunk 2 above)
Comment (samestep) [reply .github/scripts/windows_install_utils.ps1:45]: +1 on links in comments in general

**Comment by seemethere** [comment]
@seemethere has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28465272).

File: .github/workflows/pytorch-win-vs2019-cpu-py3.yml [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+      - uses: actions/setup-python@v2
+        name: Setup Python3
+        with:
+          python-version: '3.x'
+      - name: Run test scripts
+        shell: bash
+        env:
+          PYTORCH_FINAL_PACKAGE_DIR: /c/${{ github.run_id }}/build-results/
+        run: |
+            .jenkins/pytorch/win-test.sh
```
Comment (samestep) [inline .github/workflows/pytorch-win-vs2019-cpu-py3.yml:90]: nit
```suggestion
        run: |
          .jenkins/pytorch/win-test.sh
```

File: .github/workflows/pytorch-win-vs2019-cpu-py3.yml [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+      - uses: actions/upload-artifact@v2
+        name: Store PyTorch Test Reports
+        if: always()
+        with:
+          name: test-reports
+          retention-days: 30
+          if-no-files-found: error
+          path:
+            test/**/*.xml
+  # TODO: Make this into a composite step
```
Comment (samestep) [inline .github/workflows/pytorch-win-vs2019-cpu-py3.yml:100]: this isn't possible, right? since it includes several steps that aren't `run`

Comment (samestep) [reply .github/workflows/pytorch-win-vs2019-cpu-py3.yml:100]: although I guess you could turn the last four steps into a composite step

**Review by samestep — APPROVED** [review]

**Pull request closed without merging**